rfcs/rdfa-editor-stage-1.md

```diff
@@ -0,0 +1,587 @@
+the "core editing operations" needed to make useful changes to the document. Operations tend to be too
+atomic for this purpose.
+
+```typescript
+interface Mutator {
+  replaceRange(range: ModelRange, ...content: Node[]);
+  replaceNode(oldNode: Node, newNode: Node);
+  updateSelection(selection: ModelSelection);
+  selectRange(range: ModelRange);
+}
+```
+
+#### Command
+
+A command represents the highest-level primitive with which to build editing functionality. 
+Commands should be named after their intention. As certain commands only make sense in certain contexts,
+commands also carry with them a method to check whether their execution makes sense at the moment.
+
+Commands receive 2 argument objects: a config object containing the tools the command needs to interact with the document,
+and an args object containing user-defined arguments that the callsite has to provide.
+
+Commands can choose to interact with the document in 2 core ways, or a combination of both:
+- by composing other commands
+- by interacting with ranges and nodes and using the mutator to persist changes to the document
+
+```typescript
+interface CommandConfig {
+  executedBy: string;
+  controller: Controller;
+}
+
+interface Command<A> {
+  canExecute(config: CommandConfig, args: A): boolean;
+  execute(config: CommandConfig, args: A);
+}
+```
+
+### Plugins
+
+With extendability as one of the core pillars, plugin support should be considered at every level. The above
+architecture gives plugins the following capabillities:
+
+- defining new node types
+- defining new commands
+- executing commands
+- updating the vdom through a mutator (either as part of a command, or directly)
+- listening to events
+- defining and emitting new events
+- defining new UI elements and registering them for rendering
+
+```typescript
+interface EditorPlugin {
+  name: string;
+  async initialize(controller: Controller): void;
+}
+```
+#### Widgets
+
+A widget is simply a ui component which can be registered in the editor. This can be a simple menu, a single button, or even a full-blown ember application.
+At this stage, only ember-based plugins are supported, and widgets have to be ember components. There are limited locations where a widget can be registered, 
+which are specified below.
+A widget is always rendered, and as such has full control over when it is visible. 
+
+```typescript
+type WidgetLocation = "toolbar" | "sidebar" | "cursor" | "block";
+
+interface WidgetSpec {
+  componentName: string,
+  desiredLocation: WidgetLocation
+}
+```
+
+#### Adding a node type
+
+Adding a new node type is very similar to adding a command or a widget. Simply call the appropriate method in the controller, providing it with a Spec object defined below.
+The spec object demands a reader and writer to respectively deserialize and serialize the node from and to the actual DOM.
+
+```typescript
+
+interface NodeSpec {
+  name: type;
+  implementation: new () => ModelNode;
```

Comment:
What should be inside the implementation? Not 100% clear

Reply:
that's typescript syntax for "a constructor for a ModelNode"
aka you pass in the class itself